include/query/algorithm/exhaustive_taat_query.hpp

```diff
@@ -0,0 +1,138 @@
+#pragma once
+
+#include "util/intrinsics.hpp"
+#include "topk_queue.hpp"
+
+namespace pisa {
+
+// TODO: These are functions common to query processing in general.
+//       They should be moved out of this file.
+namespace query {
+
+template <typename Index, typename WandType>
+[[nodiscard]] auto cursors_with_scores(Index const& index, WandType const &wdata, term_id_vec terms)
+{
+    // TODO(michal): parametrize scorer_type; didn't do that because this might mean some more
+    //               complex refactoring I want to avoid for now.
+    using scorer_type         = bm25;
+    using cursor_type         = typename Index::document_enumerator;
+    using score_function_type = std::function<float(uint64_t, uint64_t)>;
+
+    auto query_term_freqs = query_freqs(terms);
+    std::vector<cursor_type> cursors;
+    std::vector<score_function_type> score_functions;
+    cursors.reserve(query_term_freqs.size());
+    score_functions.reserve(query_term_freqs.size());
+
+    for (auto term : query_term_freqs) {
+        auto     list     = index[term.first];
+        uint64_t num_docs = index.num_docs();
+        auto     q_weight = scorer_type::query_term_weight(term.second, list.size(), num_docs);
+        cursors.push_back(std::move(list));
+        score_functions.push_back([q_weight, &wdata](auto docid, auto freq) {
+            float norm_len = wdata.norm_len(docid);
+            return q_weight * scorer_type::doc_term_weight(freq, norm_len);
+        });
+    }
+    return std::make_pair(cursors, score_functions);
+}
+
+} // namespace query
+
+template <typename Index, typename WandType, size_t accumulator_block_size = 0>
+struct exhaustive_taat_query {
+    exhaustive_taat_query(Index const &index, WandType const &wdata, uint64_t k)
+        : m_index(index), m_wdata(wdata), m_topk(k), m_accumulators(index.num_docs())
+    {
+        static_assert(accumulator_block_size >= 0, "must be non-negative");
+        if constexpr (accumulator_block_size > 0) {
+            m_acc_block_count =
+                (m_accumulators.size() + accumulator_block_size - 1) / accumulator_block_size;
+            m_accumulators_max.resize(m_acc_block_count);
+        } else {
+            m_acc_block_count = 1;
+            m_accumulators_max.push_back(std::numeric_limits<float>::max());
+        }
+    }
+
+    uint64_t operator()(term_id_vec terms) {
+        auto cws = query::cursors_with_scores(m_index, m_wdata, terms);
+        return taat(std::move(cws.first), std::move(cws.second));
+    }
+    using score_function_type = std::function<float(uint64_t, uint64_t)>;
```

Comment:
Two things here:

1. As to the separate PR, note that this is all internal, doesn't really affect anything outside. I wanted to extract it to another function so we have logic separated. You can revert it now but it makes little sense. Exposing it as an interface should, however, be done in a separate PR, I agree.

2. As to argument types, my thinking is the following: this is always the function of a posting: (doc, freq), which is (`int`, `int`). The idea is for another function/method to return these functions. They might be, say, lambdas with captured index, which resolve the document size. It can be something completely different. The point is that by separating these two things, your processing algorithm cares not about what it must pass, it always passes the posting. Figuring out what to do with it falls onto index/scoring function, as it should.